xdsl/builder.py

```diff
@@ -23,45 +23,45 @@ class InsertPoint:
     block: Block
     """The block where the insertion point is in."""
 
-    op_before: Operation | None = field(default=None)
+    insert_before: Operation | None = field(default=None)
     """
-    The operation right before the insertion point.
-    None if the insertion point is at the end of the block.
+    The insertion point is right before this operation.
+    If the operation is None, the insertion point is at the end of the block.
     """
 
-    @overload
-    def __init__(self, op_before_or_block: Operation) -> None:
-        ...
+    @staticmethod
+    def before(op: Operation) -> InsertPoint:
+        """Gets the insertion point before an operation."""
+        if (block := op.parent_block()) is None:
+            raise ValueError("Operation insertion point must have a parent block")
+        return InsertPoint(block, op)
 
-    @overload
-    def __init__(self, op_before_or_block: Block) -> None:
-        ...
+    @staticmethod
+    def after(op: Operation) -> InsertPoint:
+        """Gets the insertion point after an operation."""
+        block = op.parent_block()
+        if block is None:
+            raise ValueError("Operation insertion point must have a parent block")
+        return InsertPoint(block, op.next_op)
 
-    @overload
-    def __init__(self, op_before_or_block: Block, op_before: Operation | None) -> None:
-        ...
+    @staticmethod
+    def at_start(block: Block) -> InsertPoint:
+        """Gets the insertion point at the start of a block."""
+        return InsertPoint(block, block.ops.first)
 
-    def __init__(
-        self, op_before_or_block: Block | Operation, op_before: Operation | None = None
-    ) -> None:
-        if isinstance(op_before_or_block, Operation):
-            block = op_before_or_block.parent_block()
-            if block is None:
-                raise ValueError("Operation insertion point must have a parent block")
-            self.block = block
-            self.op_before = op_before_or_block
-        else:
-            self.block = op_before_or_block
-            self.op_before = op_before
+    @staticmethod
+    def at_end(block: Block) -> InsertPoint:
+        """Gets the insertion point at the end of a block."""
+        return InsertPoint(block)
 
     def verify(self) -> None:
         """
         Check that the insertion point is valid.
         An insertion point can only be invalid if `op_before` is an `Operation`, and its
         parent is not `block`.
         """
-        if self.op_before is not None:
-            if self.op_before.parent is not self.block:
+        if self.insert_before is not None:
+            if self.insert_before.parent is not self.block:
                 raise ValueError("Insertion point must be in the builder's `block`")
```

Comment:
My intuition would be to make this class frozen, and move the verification to the init. The builder can make a new insertion point instead of mutating the one it has

Reply:
Fair, makes sense!